File: lib/collectors/reddit-collector.ts
export async function collectRedditPosts() {
  const subreddits = ['Forex', 'wallstreetbets', 'investing', 'economics'];
  const results = [];

  for (const subreddit of subreddits) {
    try {
      const response = await fetch(
        `https://old.reddit.com/r/${subreddit}/hot.json?limit=25`,
        {
          headers: { 
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Accept': 'application/json'
          },
          next: { revalidate: 0 }
        }
      );

      if (!response.ok) {
        console.log(`[v0] Reddit r/${subreddit} returned ${response.status}`);
        continue;
      }

      const data = await response.json();
      const posts = data.data?.children || [];

      const currencyKeywords = [
        'eur/usd', 'gbp/usd', 'usd/jpy', 'usd/chf', 'aud/usd', 'usd/cad', 'xau/usd', 'gold',
        'eurusd', 'gbpusd', 'usdjpy', 'usdchf', 'audusd', 'usdcad', 'xauusd',
        'dollar', 'euro', 'pound', 'yen', 'forex', 'currency', 'fx', 'fed', 'central bank'
      ];

      for (const post of posts) {
        const postData = post.data;
        const title = postData.title?.toLowerCase() || '';
        const selftext = postData.selftext?.toLowerCase() || '';
        const content = `${title} ${selftext}`;

        const isRelevant = currencyKeywords.some(keyword => content.includes(keyword));

        if (isRelevant && postData.ups > 5) {
          results.push({
            source_type: 'reddit',
            source_name: `r/${subreddit}`,
            content: `${postData.title}\n\n${postData.selftext || ''}`,
            url: `https://reddit.com${postData.permalink}`,
            metadata: {
              upvotes: postData.ups,
              num_comments: postData.num_comments,
              created_utc: postData.created_utc,
              author: postData.author
            }
          });
        }
      }
    } catch (error) {
      console.error(`[v0] Reddit error (r/${subreddit}):`, error);
    }
  }

  console.log(`[v0] Reddit collected ${results.length} relevant posts`);
  return results;
}

File: scripts/collectors/reddit_collector.py
"""
Reddit Collector
Collects posts and comments from forex-related subreddits
"""

import os
import json
from datetime import datetime, timedelta
from typing import List, Dict
import requests

SUBREDDITS = ['Forex', 'wallstreetbets', 'investing', 'economics', 'stocks']
CURRENCY_KEYWORDS = ['EUR/USD', 'EURUSD', 'GBP/USD', 'GBPUSD', 'USD/JPY', 'USDJPY', 
                     'XAU/USD', 'XAUUSD', 'gold', 'dollar', 'DXY']

def collect_reddit_data() -> List[Dict]:
    """Collect recent posts from forex subreddits using Reddit JSON API"""
    all_posts = []
    
    for subreddit in SUBREDDITS:
        try:
            # Reddit JSON API (no auth needed for public data)
            url = f"https://www.reddit.com/r/{subreddit}/new.json?limit=50"
            headers = {'User-Agent': 'Mozilla/5.0 (Market Intelligence Bot)'}
            
            response = requests.get(url, headers=headers, timeout=10)
            
            if response.status_code == 200:
                data = response.json()
                
                for post in data['data']['children']:
                    post_data = post['data']
                    title = post_data.get('title', '')
                    selftext = post_data.get('selftext', '')
                    content = f"{title} {selftext}".upper()
                    
                    # Check if post is currency-related
                    is_relevant = any(keyword.upper() in content for keyword in CURRENCY_KEYWORDS)
                    
                    if is_relevant:
                        post_obj = {
                            'source_platform': 'reddit',
                            'source_name': f"r/{subreddit}",
                            'content': f"Title: {title}\n\nContent: {selftext}",
                            'metadata': {
                                'url': f"https://reddit.com{post_data.get('permalink', '')}",
                                'score': post_data.get('score', 0),
                                'num_comments': post_data.get('num_comments', 0),
                                'created_utc': post_data.get('created_utc', 0),
                                'author': post_data.get('author', '[deleted]')
                            },
                            'collected_at': datetime.utcnow().isoformat()
                        }
                        all_posts.append(post_obj)
                        
        except Exception as e:
            print(f"[v0] Error collecting from r/{subreddit}: {e}")
            continue
    
    return all_posts

if __name__ == "__main__":
    posts = collect_reddit_data()
    print(json.dumps(posts, indent=2))
